src/order/lattice.lean

```diff
@@ -183,6 +263,38 @@ class semilattice_inf (α : Type u) extends has_inf α, partial_order α :=
 (inf_le_right : ∀ a b : α, a ⊓ b ≤ b)
 (le_inf : ∀ a b c : α, a ≤ b → a ≤ c → a ≤ b ⊓ c)
 
+-- TODO: deduplicate this with semilattice_sup.mk' using `order_dual` / automation
+/--
+A type with a commutative, associative and idempotent binary `inf` operation has the structure of a
+meet-semilattice.
+
+The partial order is defined so that `a ≤ b` unfolds to `a = a ⊓ b`; cf. `left_eq_inf`.
+-/
+def semilattice_inf.mk' {α : Type*} [has_inf α]
+  (inf_comm : ∀ (a b : α), a ⊓ b = b ⊓ a)
+  (inf_assoc : ∀ (a b c : α), a ⊓ b ⊓ c = a ⊓ (b ⊓ c))
+  (inf_idem : ∀ (a : α), a ⊓ a = a) : semilattice_inf α :=
+{ inf := (⊓),
+  le := λ a b, a = a ⊓ b,
+  le_refl := λ a, (inf_idem a).symm,
+  le_trans := λ a b c hab hbc,
+    begin
+      dsimp only [(≤)] at *,
+      rw [hab, inf_assoc, ←hbc],
+    end,
+  le_antisymm := λ a b hab hba,
+    begin
+      dsimp only [(≤)] at *,
+      rw [hba, inf_comm, ←hab],
+    end,
+  inf_le_left  := λ a b, show a ⊓ b = a ⊓ b ⊓ a, by rw [inf_comm, inf_assoc, inf_idem],
+  inf_le_right := λ a b, show a ⊓ b = a ⊓ b ⊓ b, by rw [inf_assoc, inf_idem],
+  le_inf := λ a b c hac hbc,
+    begin
+      dsimp only [(≤), preorder.le] at *,
+      rwa [←inf_assoc, ←hac],
+    end }
```

Comment:
I don't know if this is Kosher, but:
```suggestion
begin
  haveI : semilattice_sup (order_dual α) := semilattice_sup.mk' inf_comm inf_assoc inf_idem,
  haveI i := order_dual.semilattice_inf (order_dual α),
  exact i,
end
```

Reply:
Strictly I think you ought to have a lemma like:
```
theorem preorder.dual_dual (α : Type*) [H : preorder α] :
  order_dual.preorder (order_dual α) = H :=
preorder.ext $ λ _ _, iff.rfl
```
but for semilattice_inf, and then use that on the last line.

Reply:
Not sure what you mean with the last suggestion. It might be easier if you push your changes to this branch?

Reply:
I think those lemmas are worth having whether you use them or not, but maybe @gebner or whoever it was that was worrying about defeq abuse can chime in about how to use them here.